srcdeps-core/src/main/java/org/l2x6/srcdeps/core/fs/PathLocker.java

```diff
@@ -60,65 +97,129 @@ private static void close(RandomAccessFile lockFile, Path lockFilePath, Reentran
+     * </tr>
+
+     * <tr>
+     * <td>yes</td>
+     * <td>no</td>
+     * <td>-</td>
+     * <td>{@link CannotAcquireLockException} thrown immediately</td>
+     * </tr>
+     *
+     * <tr>
+     * <td>no</td>
+     * <td>-</td>
+     * <td>yes</td>
+     * <td>{@link CannotAcquireLockException} thrown immediately</td>
+     * </tr>
+     *
+     * <tr>
+     * <td>no</td>
+     * <td>-</td>
+     * <td>no</td>
+     * <td>Locked immediately</td>
+     * </tr>
+     * </table>
+     *
      * <p>
      * The returned {@link PathLock} should be released using its {@link Closeable#close()} method.
      *
      * @param path
      *            the {@link Path} to lock
+     * @param pathMetadata
+     *            a metadata associated with the given {@code path}
      * @return a {@link PathLock} whose holder is guaranteed to have an exclusive access to {@link PathLock#getPath()}
      * @throws IOException
      *             if the given {@code path} cannot be created as a directory
      * @throws CannotAcquireLockException
      *             if the lock cannot be acquired immediately
-     * @throws LockerDisposedException
      */
-    public PathLock tryLockDirectory(Path path) throws IOException, CannotAcquireLockException {
+    public PathLock lockDirectory(Path path, M pathMetadata) throws IOException, CannotAcquireLockException {
         SrcdepsCoreUtils.ensureDirectoryExists(path);
-        Path lockFilePath = path.resolveSibling(path.getName(path.getNameCount() - 1) + ".lock");
         // resolve(String.valueOf(i) + ".lock");
-        final ReentrantLock newLock = new ReentrantLock();
-        final ReentrantLock oldLock = locks.putIfAbsent(lockFilePath, newLock);
-        final ReentrantLock threadLevelLock = oldLock == null ? newLock : oldLock;
-        if (threadLevelLock.tryLock()) {
-            log.debug("Locked on thread level {}", path);
-            RandomAccessFile lockFile = null;
-            try {
-                lockFile = new RandomAccessFile(lockFilePath.toFile(), "rw");
-                FileLock fsLock = lockFile.getChannel().tryLock();
-                log.debug("Locked on FS {} with lock {}", path, fsLock);
-                if (fsLock == null) {
-                    throw new CannotAcquireLockException(
-                            String.format("Could not acquire filesystem level lock on [%s]", lockFilePath));
-                } else {
-                    return new PathLock(path, lockFile, lockFilePath, threadLevelLock);
-                }
-            } catch (CannotAcquireLockException e) {
-                close(lockFile, lockFilePath, threadLevelLock);
-                throw e;
-            } catch (OverlappingFileLockException e) {
+        final LockMetadataPair<M> newPair = new LockMetadataPair<M>(new ReentrantLock(), pathMetadata);
+        final LockMetadataPair<M> oldPair = locks.putIfAbsent(path, newPair);
+        final LockMetadataPair<M> mdPair = oldPair == null ? newPair : oldPair;
```

Comment:
> I'd do it wrapping in a conditional state

I was considering it, but then I decided that because `newPair` will be the winner in most cases (several threads requesting the same thing at once will be rather rare) the additional `locks.get()` is not justified enough.